src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py

```diff
@@ -389,262 +495,105 @@ def push_average(self, metrics: List[Tuple[str, Receiver[Sample], bool]]) -> Non
-    ) -> _GenericHOFormulaBuilder:
-        """Return a formula builder that subtracts (data in) `other` from `self`.
-
-        Args:
-            other: A formula receiver, or a formula builder instance corresponding to a
-                sub-expression.
-
-        Returns:
-            A formula builder that can take further expressions, or can be built
-                into a formula engine.
-        """
-        return self.BuilderType(self) - other
-
-    def __mul__(
-        self, other: _BaseFormulaReceiver | _GenericHOFormulaBuilder
-    ) -> _GenericHOFormulaBuilder:
-        """Return a formula builder that multiplies (data in) `self` with `other`.
-
-        Args:
-            other: A formula receiver, or a formula builder instance corresponding to a
-                sub-expression.
-
-        Returns:
-            A formula builder that can take further expressions, or can be built
-                into a formula engine.
-        """
-        return self.BuilderType(self) * other
+        while self._build_stack:
+            self._steps.append(self._build_stack.pop())
 
-    def __truediv__(
-        self, other: _BaseFormulaReceiver | _GenericHOFormulaBuilder
-    ) -> _GenericHOFormulaBuilder:
-        """Return a formula builder that divides (data in) `self` by `other`.
+        return self._steps, self._metric_fetchers
 
-        Args:
-            other: A formula receiver, or a formula builder instance corresponding to a
-                sub-expression.
+    def build(self) -> FormulaEngine:
+        """Create a formula engine with the steps and fetchers that have been pushed.
 
         Returns:
-            A formula builder that can take further expressions, or can be built
-                into a formula engine.
+            A `FormulaEngine` instance.
         """
-        return self.BuilderType(self) / other
+        self.finalize()
+        return FormulaEngine(self)
 
 
-class _BaseHOFormulaBuilder(
-    ABC, Generic[_GenericFormulaReceiver, _GenericSample, _GenericEngine]
-):
+class _BaseHOFormulaBuilder(ABC):
     """Provides a way to build formulas from the outputs of other formulas."""
 
-    def __init__(self, recv: _GenericFormulaReceiver) -> None:
+    def __init__(self, engine: FormulaEngine | FormulaEngine3Phase) -> None:
         """Create a `GenericHigherOrderFormulaBuilder` instance.
 
         Args:
-            recv: A first input stream to create a builder with, so that python
+            engine: A first input stream to create a builder with, so that python
                 operators `+, -, *, /` can be used directly on newly created instances.
         """
-        self._steps: deque[tuple[TokenType, _GenericFormulaReceiver | str]] = deque()
-        self._steps.append((TokenType.COMPONENT_METRIC, recv.clone()))
-        recv._deactivate()  # pylint: disable=protected-access
-        self._engine = None
+        self._engine: FormulaEngine | FormulaEngine3Phase
```

Comment:
Should `self._engine` also be None and set to None?

Reply:
Actually I didn't spot any reason in the changes to keep this attribute  `self._engine`. Can you please double-check if it is still needed?

Reply:
already addressed